autoload/ale/filerename.vim

```diff
@@ -0,0 +1,127 @@
+" Author: Dalius Dobravolskas <dalius.dobravolskas@gmail.com>
+" Description: Rename file support for tsserver
+
+let s:filerename_map = {}
+
+" Used to get the rename map in tests.
+function! ale#filerename#GetMap() abort
+    return deepcopy(s:filerename_map)
+endfunction
+
+" Used to set the rename map in tests.
+function! ale#filerename#SetMap(map) abort
+    let s:filerename_map = a:map
+endfunction
+
+function! ale#filerename#ClearLSPData() abort
+    let s:filerename_map = {}
+endfunction
+
+function! s:message(message) abort
+    call ale#util#Execute('echom ' . string(a:message))
+endfunction
+
+function! ale#filerename#HandleTSServerResponse(conn_id, response) abort
+    if get(a:response, 'command', '') isnot# 'getEditsForFileRename'
+        return
+    endif
+
+    if !has_key(s:filerename_map, a:response.request_seq)
+        return
+    endif
+
+    let l:options = remove(s:filerename_map, a:response.request_seq)
+
+    let l:old_name = l:options.old_name
+    let l:new_name = l:options.new_name
+
+    if get(a:response, 'success', v:false) is v:false
+        let l:message = get(a:response, 'message', 'unknown')
+        call s:message('Error renaming file "' . l:old_name . '" to "' . l:new_name
+        \ . '". Reason: ' . l:message)
+
+        return
+    endif
+
+    let l:changes = a:response.body
+
+    if empty(l:changes)
+        call s:message('No changes while renaming "' . l:old_name . '" to "' . l:new_name . '"')
+    else
+        call ale#code_action#HandleCodeAction(
+        \   {
+        \       'description': 'filerename',
+        \       'changes': l:changes,
+        \   },
+        \   {
+        \       'should_save': 1,
+        \   },
+        \)
+    endif
+
+    silent! noautocmd execute 'saveas ' . l:new_name
+    call delete(l:old_name)
+endfunction
+
+function! s:OnReady(options, linter, lsp_details) abort
+    let l:id = a:lsp_details.connection_id
+
+    if !ale#lsp#HasCapability(l:id, 'filerename')
```

Comment:
@daliusd add the filerename capability to the default capabilities in the ale#lsp#Register() method. If not I get an exception thrown when checking for that capability here when and LSP that is not tsserver is active.

```
New file name: /home/ryujin/Projects/HelloWorld/python/hello-plain/hello.pyError detected while processing function ale#filerename#Execute[26]..<SNR>202_ExecuteFileRename[4]..ale#lsp_linter#StartLSP[26]..<SNR>190_StartIfExecutable[15]..<SNR>190_StartWithCommand[13]..<SNR>190_StartLSP[49]..ale#lsp#On
Init[8]..<lambda>24[1]..ale#lsp_linter#OnInit[45]..<SNR>202_OnReady[3]..ale#lsp#HasCapability:
line    8:
E605: Exception not caught: Invalid capability filerename
Error detected while processing function ale#filerename#Execute[26]..<SNR>202_ExecuteFileRename[4]..ale#lsp_linter#StartLSP[26]..<SNR>190_StartIfExecutable[15]..<SNR>190_StartWithCommand[13]..<SNR>190_StartLSP[49]..ale#lsp#OnInit[8]..<lambda>24[1]..ale#lsp_linter#OnInit[45]..<SNR>202_OnReady:
line    3:
E171: Missing :endif
Error detected while processing function ale#filerename#Execute[26]..<SNR>202_ExecuteFileRename[4]..ale#lsp_linter#StartLSP[26]..<SNR>190_StartIfExecutable[15]..<SNR>190_StartWithCommand[13]..<SNR>190_StartLSP[49]..ale#lsp#OnInit:
line    8:
E171: Missing :endif
Error detected while processing function ale#filerename#Execute:
line   26:
E170: Missing :endfor
```

Reply:
Added. I have not noticed this error before, tested with `pyright` - must be good now.